File: quiz_app_flutter/lib/presentation/dashboard/viewmodel/dashboard_viewmodel.dart
import 'dart:async';

import 'package:connectivity_plus/connectivity_plus.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:quiz_app_flutter/di/flutter_riverpod.dart';
import 'package:quiz_app_flutter/domain/entities/dashboard/dashboard_data.dart';
import 'package:quiz_app_flutter/presentation/dashboard/intent/dashboard_intent.dart';
import 'package:quiz_app_flutter/presentation/dashboard/intent/dashboard_nav_event.dart';
import 'package:quiz_app_flutter/presentation/utils/viewmodel/base_viewModel.dart';

class DashboardViewModel extends BaseViewModel<DashboardData> {
  final Ref _ref;

  // StreamController for navigation events
  final _navigationEventsController =
      StreamController<DashboardNavEvent>.broadcast();
  Stream<DashboardNavEvent> get navigationEvents =>
      _navigationEventsController.stream;

  DashboardViewModel(this._ref) : super() {
    setLoading();
    // automatically load dashboard when ViewModel is created
    handleIntent(LoadDashboard());
  }

  Future<void> handleIntent(DashboardIntent intent) async {
    if (intent is LoadDashboard) {
      await loadDashboard();
    } else if (intent is NavigateToQuizSets) {
      _navigationEventsController.add(
        DashboardNavEventNavigateToQuizSets(intent.item),
      );
    }
  }

  Future<void> loadDashboard() async {
    setLoading();

    try {
      final connectivityResult = await Connectivity().checkConnectivity();
      if (connectivityResult == ConnectivityResult.none) {
        setError('No internet connection');
        return;
      }

      final useCase = await _ref.read(getDashboardDataUseCaseProvider.future);

      useCase().listen((resource) {
        resource.when(
          success: (data) => setSuccess(data),
          failure: (e) => setError(e.toString()),
          loading: () => setLoading(),
        );
      });
    } catch (e) {
      setError(e.toString());
    }
  }

  @override
  void dispose() {
    _navigationEventsController.close();
    super.dispose();
  }
}

File: quiz_app_kmp/composeApp/src/commonMain/kotlin/com/tnm/quizappkmp/presentation/dashboard/viewmodel/DashboardViewModel.kt
package com.tnm.quizappkmp.presentation.dashboard.viewmodel

import androidx.lifecycle.viewModelScope
import com.tnm.quizappkmp.domain.model.Resource
import com.tnm.quizappkmp.domain.model.dashboard.DashboardData
import com.tnm.quizappkmp.domain.usecase.dashboard.GetDashboardDataUseCase
import com.tnm.quizappkmp.presentation.dashboard.intent.DashboardIntent
import com.tnm.quizappkmp.presentation.dashboard.intent.DashboardNavEvent
import com.tnm.quizappkmp.presentation.utils.viewmodel.BaseViewModel
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.onStart
import kotlinx.coroutines.launch

class DashboardViewModel(val getDashboardDataUseCase: GetDashboardDataUseCase) : BaseViewModel<DashboardData>() {
    private val _navigationEvents = MutableSharedFlow<DashboardNavEvent>()
    val navigationEvents = _navigationEvents.asSharedFlow()
    fun handleIntent(intent: DashboardIntent) {
        when (intent) {
            is DashboardIntent.LoadDashboard -> loadDashboard()
            is DashboardIntent.NavigateToQuizSets -> navigateToDetails(intent.item)
        }
    }

    private fun loadDashboard() {
        viewModelScope.launch {
            getDashboardDataUseCase()
                .onStart {
                    setLoading()
                }
                .catch { e ->
                    setError(e.message.orEmpty())
                }
                .collect { dashboard ->
                    when (dashboard) {
                        is Resource.Failure -> setError(dashboard.error.message.orEmpty())
                        is Resource.Success -> {
                            setSuccess(dashboard.data)
                        }
                    }

                }
        }
    }

    private fun navigateToDetails(item: DashboardData.Item) {
        viewModelScope.launch {
            _navigationEvents.emit(DashboardNavEvent.NavigateToQuizSets(item))
        }
    }

}